frontend/packages/kubevirt-plugin/src/components/modals/scheduling-modals/shared/hoc.ts

```diff
@@ -0,0 +1,97 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { FirehoseResult } from '@console/internal/components/utils';
+import { getLabels } from '@console/shared';
+import { NodeModel } from '@console/internal/models';
+import { getLoadedData, isLoaded, getLoadError } from '../../../../utils';
+import { VMLikeEntityKind } from '../../../../types/vmLike';
+import { NodeKind } from '../../../../../../../public/module/k8s';
+import { IDLabel } from './types';
+
+export const useNodeQualifier = <T extends IDLabel>(
+  initialLabels: T[],
+  nodes: FirehoseResult<VMLikeEntityKind[]>,
+): [NodeKind[], string] => {
+  const loadError = getLoadError(nodes, NodeModel);
+  const loadedNodes = getLoadedData(nodes, []);
+  const [qualifiedNodes, setQualifiedNodes] = React.useState([]);
+
+  React.useEffect(() => {
+    if (isLoaded(nodes)) {
+      const labels = initialLabels.filter(({ key }) => !!key);
+      const newNodes = [];
+      loadedNodes.forEach((node) => {
+        const nodeLabels = getLabels(node);
+        if (nodeLabels && labels.every(({ key, value }) => nodeLabels[key] === value)) {
+          newNodes.push(node);
+        }
+      });
+      setQualifiedNodes(newNodes);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [initialLabels, nodes]);
+
+  return [qualifiedNodes, loadError];
+};
+
+export const useLabels = <T extends IDLabel>(
+  initialLabels: T[],
+  defaultLabel: T,
+): [
+  T[], // labels
+  React.Dispatch<React.SetStateAction<T[]>>, // setLabels()
+  () => void, // addLabel()
+  (newLabel: T) => void, // changeLabel()
+  (idToDelete: number) => void, // deleteLabel()
+] => {
+  const [labels, setLabels] = React.useState<T[]>(initialLabels);
+
+  const onLabelAdd = () => {
+    const id = labels[labels.length - 1]?.id + 1 || 0;
+    setLabels([...labels, { ...defaultLabel, id }]);
+  };
+
+  const onLabelChange = (newLabel: T) =>
+    setLabels(
+      labels.map((label) => {
+        if (label.id === newLabel.id) {
+          return newLabel;
+        }
+        return label;
+      }),
+    );
+
+  const onLabelDelete = (idToDelete: number) => {
+    setLabels(labels.filter(({ id }) => id !== idToDelete));
+  };
+
+  return [labels, setLabels, onLabelAdd, onLabelChange, onLabelDelete];
+};
+
+export const useCollisionChecker = <T>(
+  vmLikeEntity: VMLikeEntityKind,
+  vmLikeEntityLoading: FirehoseResult<VMLikeEntityKind>,
+  getEntity: (vm: VMLikeEntityKind) => T,
+  setEntity: React.Dispatch<React.SetStateAction<T>>,
```

Comment:
it does too many things. setEntity should be a responsibility of the caller of reload